src/advertise.rs

```diff
@@ -0,0 +1,271 @@
+        let domain = format!("{service_type}._tcp.local.");
+        let mut props = HashMap::new();
+
+        props.insert("td".to_string(), path);
+        props.insert("type".to_string(), ty.to_dns_type().to_string());
+
+        let service = ServiceInfo::new(
+            &domain,
+            name.as_ref(),
+            &hostname,
+            ips.as_slice(),
+            port,
+            Some(props),
+        )?;
+
+        mdns.register(service)?;
+
+        Ok(())
+    }
+}
+
+impl Advertiser {
+    /// Create a new service advertiser
+    pub fn new() -> Result<Self> {
+        let mdns = ServiceDaemon::new()?;
+
+        let mut hostname = hostname::get()?.to_string_lossy().to_string();
+        if !hostname.ends_with(".local") {
+            hostname.push_str(".local");
+        }
+        let ips = if_addrs::get_if_addrs()?
+            .iter()
+            .filter_map(|iface| {
+                let ip = iface.ip();
+                match ip {
+                    IpAddr::V4(ip) if !iface.is_loopback() => Some(ip),
+                    _ => None,
+                }
+            })
+            .collect();
+
+        let sa = Self {
+            mdns,
+            ips,
+            hostname,
+        };
+
+        Ok(sa)
+    }
+
+    /// Register and Advertise a new service
+    pub fn add_service(&self, name: impl AsRef<str>) -> ServiceBuilder {
+        ServiceBuilder::new(self, name)
+    }
+}
+
+#[cfg(all(test, not(miri)))]
+mod test {
+    use super::*;
+
+    use mdns_sd::{ServiceEvent::*, ServiceInfo};
+    use std::time::Duration;
+
+    #[test]
+    fn set_hostname() {
+        test_feature(
+            "TestLampHostname",
+            "_wot._tcp.local.",
+            |b| b.hostname("testhost"),
+            |info| {
+                let props = info.get_properties();
+                assert_eq!(props["td"], WELL_KNOWN);
+                assert_eq!(props["type"], "Thing");
+                assert_eq!(info.get_hostname(), "testhost.");
+            },
+        );
+    }
+
+    #[test]
+    fn set_path() {
+        test_feature(
+            "TestLampPath",
+            "_wot._tcp.local.",
+            |b| b.path("/test/path"),
+            |info| {
+                let props = info.get_properties();
+                assert_eq!(props["td"], "/test/path");
+                assert_eq!(props["type"], "Thing");
+            },
+        );
+    }
+
+    #[test]
+    fn set_type() {
+        test_feature(
+            "TestDirectory",
+            "_directory._sub._wot._tcp.local.",
+            |b| b.thing_type(ThingType::Directory),
+            |info| {
+                let props = info.get_properties();
+                assert_eq!(props["td"], WELL_KNOWN);
+                assert_eq!(props["type"], "Directory");
+            },
+        );
+    }
+
+    fn test_feature<F>(name: &str, browse: &str, build: F, check: fn(ServiceInfo))
+    where
+        F: for<'b> Fn(ServiceBuilder<'b>) -> ServiceBuilder<'b>,
```

Comment:
FYI: I think that in this case it should not be necessary so use a HRTB. Annotating `name: `&'a str` and using `'a` for the bound should be sufficient (don't bother to change the code, it's fine)